CTP-4793 moderated marking behat tests

opitz opened the pull request:
Added Behat tests for moderated marking according to https://liveuclac-my.sharepoint.com/:x:/g/personal/dtnvgal_ucl_ac_uk/ESRtBFnS-xVPmFJl-YynDLsBRKCXh-zHVUgfq0ydIGWSkw?e=FrVGTS&wdOrigin=TEAMS-MAGLEV.p2p_ns.rwc&wdExp=TEAMS-TREATMENT&wdhostclicktime=1763465662042&web=1

Tests involved Turnitin are dealt with in a separate story.

opitz committed:
Scenario: Create coursework assignment with double marking

opitz committed:
Scenario: Add markers

opitz committed:
Scenario: Allocate markers

opitz committed:
Scenario: Check anonymity

opitz committed:
cleanup

opitz committed:
using new behat step 'And there is a double-blind marking coursework'

opitz committed:
Scenario: Add extension to a student

opitz committed:
Scenario: Student can submit a PDF file
Scenario: Student with extension can submit after deadline w/o being late
Scenario: Student has no extension so submission is late
Added custom Behat step:
- Set an extension deadline for a student in a coursework.
- For matching a late submitted date ignoring the time part

opitz committed:
No 'formative or summative' setting

opitz committed:
addressed GHA issues

opitz committed:
added Behat steps to create a submission and/or finalise it for a given student

opitz committed:
Scenario: Manager can submit on behalf of students.

opitz committed:
added behat step i_assign_user_as_role_for_student_in_coursework()

opitz committed:
added Behat step i_click_add_mark_for_marker_in_row()
added Behat step i_should_see_mark_in_row()

opitz committed:
added test PDF files, addressed Code Checker issues

opitz committed:
Scenario: Mark the assignments

opitz committed:
addressed more Code Checker issues

opitz committed:
defining courseworkmarker role capabilities

opitz committed:
'percentage distance' => 'Percentage distance'

opitz committed:
defining correct roles for courseworkexamoffice, courseworkmarker and courseworkmoderator

opitz committed:
Scenario: Moderate the assessment

opitz committed:
Scenario: Check moderation

opitz committed:
Scenario: Release the grades

opitz committed:
Scenario: Student 1 sees the released grades

opitz committed:
Scenario: Student 2 sees disagreed released grades

opitz committed:
Scenario: Student 3 does not see unmoderated released grades

opitz committed:
Scenario: Check moderation form

andrewhancox reviewed (changes requested):
I'm not sure how other people feel but I really don't think we should be creating any more custom behat steps unless we can possibly avoid it - if anything we should be going the other way.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3680,4 +3704,58 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+            $timestamp = $datestr;
+        } else {
+            $timestamp = strtotime($datestr);
+        }
```
We could just use the ternary operator where we use this value:

```
        $record->extended_deadline = is_int($datestr) ? $datestr : strtotime($datestr);
```

opitz replied:
We do now. Thanks!

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3823,4 +3813,73 @@ public function set_extension_for_user($fullname, $cwname, $datestr) {
+     *
+     * @param $fullname
+     * @return mixed|stdClass
+     * @throws dml_exception
```
This function may throw an exception but not this class.

opitz replied:
refactored.

opitz committed:
removed test case where student should not see unmoderated feedback

opitz committed:
addressed code review issues by copilot

opitz committed:
addressed code review issues by copilot2

opitz committed:
refactored assigning role capabilities in Behat tests

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1591,6 +1591,28 @@ public function there_is_a_coursework() {
+        $coursework->filetypes = "pdf";
+
+        $this->coursework = coursework::find($generator->create_instance($coursework)->id);
+    }
```
This is similar to `there_is_a_blind_marking_moderation_coursework()`, could we combine them:

```
    /**
     ⋮
     * @Given /^there is a (double-)?blind marking( moderated)? coursework$/
     */
    public function there_is_a_blind_marking_coursework(bool $double = false, bool $moderated = false) {

        /**
         * @var $generator mod_coursework_generator
         */
        $generator = testing_util::get_data_generator()->get_plugin_generator('mod_coursework');
    
        $coursework = new stdClass();
        $coursework->course = $this->course;
        $coursework->startdate = time();
        $coursework->deadline = strtotime('+15 minutes');
        $coursework->numberofmarkers = ($double ? 2 : 1);
        $coursework->blindmarking = true;
        $coursework->allocationenabled = true;
        $coursework->extensionsenabled = true;
        $coursework->allowlatesubmissions = true;
        $coursework->moderationagreementenabled = $moderated;
        $coursework->filetypes = "pdf";

        $this->coursework = coursework::find($generator->create_instance($coursework)->id);
    }
```

for steps `there is a double-blind marking coursework` and `there is a blind marking moderated coursework` (`moderation` changed to `moderated`).

opitz replied:
Thank you very much for this.

opitz commented on tests/generator/behat_mod_coursework_generator.php:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Behat data generator for mod_assign.
+ *
+ * @package   mod_assign
```
file behat_mod_coursewrk_generator.php removed for now.

opitz commented on tests/generator/behat_mod_coursework_generator.php:
```diff
@@ -0,0 +1,165 @@
+                'singular' => 'extension',
+                'datagenerator' => 'extension',
+                'required' => ['assign', 'user', 'extensionduedate'],
+                'switchids' => ['assign' => 'cmid', 'user' => 'userid'],
```
file behat_mod_coursewrk_generator.php removed for now.

opitz commented on tests/generator/behat_mod_coursework_generator.php:
```diff
@@ -0,0 +1,165 @@
+     * @param TableNode $data
+     *
+     */
+    /**
```
file behat_mod_coursewrk_generator.php removed for now.

opitz committed:
removed behat generator file for now, addressed copilot issues

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3823,4 +3813,73 @@ public function set_extension_for_user($fullname, $cwname, $datestr) {
+        } else {
+            $DB->insert_record('coursework_allocation_pairs', $record);
+        }
+    }
```
If we change this function to:

```
    /**
     * @Given /^I allocate the following markers:$/
     *
     * @param TableNode $allocations Students and their markers.
     */
    public function i_allocate_the_following_markers(TableNode $allocations) {
        global $DB;

        $datahash = $allocations->getHash();

        foreach ($datahash as $allocate) {
            $stages = ['assessor_1'];
            $stages[] = isset($allocate['moderator']) ? 'moderator' : 'assessor_2';
            $student = $this->get_user_from_username($allocate['student']);
            foreach ($stages as $stage) {
                $marker = $this->get_user_from_username($allocate[$stage]);

                $record = $DB->get_record('coursework_allocation_pairs', [
                    'courseworkid' => $this->coursework->id,
                    'allocatableid' => $student->id,
                    'allocatableuser' => $student->id,
                    'stageidentifier' => $stage,
                    'allocatabletype' => 'user',
                ]);
                if ($record) {
                    $record->assessorid = $marker->id;
                    $record->ismanual = 1;
                    $DB->update_record('coursework_allocation_pairs', $record);
                } else {
                    $record = new stdClass();
                    $record->courseworkid = $this->coursework->id;
                    $record->allocatableid = $student->id;
                    $record->allocatableuser = $student->id;
                    $record->assessorid = $marker->id;
                    $record->stageidentifier = $stage;
                    $record->allocatabletype = 'user';
                    $record->ismanual = 1;
                    $DB->insert_record('coursework_allocation_pairs', $record);
                }
            }
        }
    }
```

then in `double_marking_blind.feature` we can change:

```
    And I assign user "Marker 1" as "Assessor 1" for "Student 1" in coursework "Coursework 1"
    And I assign user "Marker 2" as "Assessor 2" for "Student 1" in coursework "Coursework 1"
    And I assign user "Marker 1" as "Assessor 1" for "Student 2" in coursework "Coursework 1"
    And I assign user "Marker 2" as "Assessor 2" for "Student 2" in coursework "Coursework 1"
    And I assign user "Marker 1" as "Assessor 1" for "Student 3" in coursework "Coursework 1"
    And I assign user "Marker 2" as "Assessor 2" for "Student 3" in coursework "Coursework 1"
```

to:

```
    And I allocate the following markers:
      | student   | assessor_1 | assessor_2 |
      | Student 1 | Marker 1   | Marker 2   |
      | Student 2 | Marker 1   | Marker 2   |
      | Student 3 | Marker 1   | Marker 2   |
```

and change:

```
    And I assign user "Marker 1" as "Assessor 1" for "Student 1" in coursework "Coursework 1"
    And I assign user "Moderator 1" as "Moderator" for "Student 1" in coursework "Coursework 1"
    And I assign user "Marker 1" as "Assessor 1" for "Student 2" in coursework "Coursework 1"
    And I assign user "Moderator 1" as "Moderator" for "Student 2" in coursework "Coursework 1"
    And I assign user "Marker 2" as "Assessor 1" for "Student 3" in coursework "Coursework 1"
    And I assign user "Moderator 1" as "Moderator" for "Student 3" in coursework "Coursework 1"
```

to:

```
    And I allocate the following markers:
      | student   | assessor_1 | moderator   |
      | Student 1 | Marker 1   | Moderator 1 |
      | Student 2 | Marker 1   | Moderator 1 |
      | Student 3 | Marker 2   | Moderator 1 |
```

opitz replied:
Thanks!!!

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3882,4 +3882,40 @@ private function get_user_from_username($fullname) {
+            "//a[@data-mark-action='addfeedback']";
+
+        $this->getSession()->getPage()->find('xpath', $xpath)->click();
+    }
```
For some reason I don't like this function. _Add mark_ is a link not a button.

And since markers are allocated isn't there a single button in each row making `marker "Marker 1"` superfluous?

We could remove this function and edit `double_marking_blind.feature` changing:

```
    When I click the "Add mark" button for marker "Marker 1" in row "1"
    ⋮
    When I click the "Add mark" button for marker "Marker 1" in row "2"
    ⋮
    When I click the "Add mark" button for marker "Marker 1" in row "3"
```

to:

```
    And I click on "Add mark" "link" in the "(//tr[contains(@class,'mod-coursework-submissions-row')])[1]" "xpath_element"
    ⋮
    And I click on "Add mark" "link" in the "(//tr[contains(@class,'mod-coursework-submissions-row')])[2]" "xpath_element"
    ⋮
    And I click on "Add mark" "link" in the "(//tr[contains(@class,'mod-coursework-submissions-row')])[3]" "xpath_element"
```

opitz replied:
refactored

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -299,3 +299,62 @@ Feature: Double marking - blind
+    Then I should see "Submissions"
+    Then I should see the mark "70" in row "1"
+    Then I should see the mark "70" in row "2"
+    Then I should see the mark "70" in row "3"
```
If we change the marks to 71, 72 and 73 then we could have;

```
    Then I should see "Submissions"
    And "71" "link" should exist
    And "72" "link" should exist
    And "73" "link" should exist
```

Then we could remove `i_should_see_mark_in_row()`.

opitz replied:
clever! refactored.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3963,4 +3963,131 @@ public function role_has_capabilities($rolename, \Behat\Gherkin\Node\TableNode $
+        // Resolve coursework module by name.
+        if (!$coursework = $DB->get_record('coursework', ['name' => $courseworkname])) {
+            throw new ExpectationException("Coursework '$courseworkname' not found", $this->getSession());
+        }
```
We can use `$this->coursework->id` so we can drop the `$courseworkname` entry parameter and make the step:

```
    And the submission from "Student 1" is marked by "Marker 1" with:
```

opitz replied:
refactored

opitz commented on tests/generator/behat_mod_coursework_generator.php:
```diff
@@ -0,0 +1,165 @@
+                'switchids' => ['assign' => 'cmid', 'user' => 'userid'],
+            ],
+        ];
+    }
```
file behat_mod_coursewrk_generator.php removed for now.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4090,4 +4052,188 @@ public function allow_role_to_assign_role(string $fromrolename, string $torolena
+     *
+     * @Given /^the student "(?P<studentname>(?:[^"]|\\")*)" has a submission for coursework "(?P<courseworkname>[^"]+)"$/
+     */
+    public function student_has_a_submission_for_coursework($studentfullname, $courseworkname) {
```
For CTP-5238 `named_student_has_a_submission()` was added which seems very similar. Can we combine them into a single function so we can replace:

```
    And the student "Student 1" has a submission for coursework "Coursework 1"
    And the submission for "Student 1" in "Coursework 1" is finalised
```

with:

```
    And the student called "Student 1" has a finalised submission
```

opitz replied:
Thank you for this. Refactored.

opitz committed:
addressed (some) code checker issues

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4090,4 +4052,188 @@ public function allow_role_to_assign_role(string $fromrolename, string $torolena
+     *
+     * @Given /^the submission for "(?P<studentname>(?:[^"]|\\")*)" in "(?P<courseworkname>[^"]+)" is (not )?finalised$/
+     */
+    public function submission_for_student_in_coursework_is_finalised($studentfullname, $courseworkname, $negate = false) {
```
If we use `named_student_has_a_submission()` we can pass the finalised status in that step so won't need this function.

opitz replied:
Nice, refactored.
Also amended named_student_has_a_submission() to allow full names ('Firstname Lastname') as well.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4090,4 +4052,188 @@ public function allow_role_to_assign_role(string $fromrolename, string $torolena
+     *
+     * @Given /^the feedback for "(?P<studentfullname>[^"]+)" by "(?P<markerfullname>[^"]+)" in "(?P<courseworkname>[^"]+)" is finalised$/
+     */
+    public function finalise_feedback($studentfullname, $markerfullname, $courseworkname) {
```
This function is only used by a commented out step:

```
#    Alternative separate step to finalise a feedback.
#    And the feedback for "Student 1" by "Marker 1" in "Coursework 1" is finalised
```

Do we need it or not?

opitz replied:
Removed.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4090,4 +4052,188 @@ public function allow_role_to_assign_role(string $fromrolename, string $torolena
+        }
+
+        // Click it.
+        $link->click();
```
We could do this with XPaths and make it shorter:

```
        $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//a[text()='$linktext']";

        $element = $this->getSession()->getPage()->find('xpath', $xpath);

        if (!$element) {
            $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//button[text()='$linktext']";
        }

        if (!$element) {
            throw new ExpectationException(
                "Could not find a link or button '{$linktext}' in row {$rownumber}.",
                $this->getSession()
            );
        }

        // Click it.
        $element->click();
```

opitz replied:
That is actually like a lecture for me, thanks for that!
Refactored (and added a missing find() after pointing the path to a button)

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4090,4 +4052,188 @@ public function allow_role_to_assign_role(string $fromrolename, string $torolena
+        if (strpos($row->getText(), $text) === false) {
+            throw new Exception("The text '{$text}' was not found in row {$rownumber}.\nRow contents: " . $row->getText());
+        }
+    }
```
We could chop this from 31 to 7 lines:

```
        $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//*[contains(text(), '$text')]";
        $element = $this->getSession()->getPage()->find('xpath', $xpath);
    
        if ($not && $element) {
            throw new Exception("'$text' text was found in row $rownumber.");
        } else if (!$element) {
            throw new Exception("'$text' was not found in row $rownumber.");
        } 
```

opitz replied:
I REALLY need to look more into the xpath usage. Thanks again, refactored!

opitz replied:
@leonstr I actually have to revoke this changes. It seems to work in most cases but not for scenario: Release the grades. 
There it fails with 'And I should see "Released" in row "1"'. I guess the xpath may need to be amended to be able to cover this a s well? Could you help with that, please? 
For now returning to my version as it works...

leonstr replied:
This should work (I tested it):

```
        $xpath = "//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr[$rownumber]//*[text()[contains(.,'$text')]]";
        $element = $this->getSession()->getPage()->find('xpath', $xpath);

        if ($notvisible && $element) {
            ⋮
        } else if (!$element) {
            ⋮
        }
```

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4236,4 +4236,87 @@ public function i_should_see_text_in_row($text, $rownumber) {
+        // Resolve coursework module by name.
+        if (!$coursework = $DB->get_record('coursework', ['name' => $courseworkname])) {
+            throw new ExpectationException("Coursework '$courseworkname' not found", $this->getSession());
+        }
```
We can use `$this->coursework->id` and shorten this step to `And the submission from "Student 1" is moderated by "Moderator 1" with:`.

opitz replied:
refactored

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -4330,4 +4330,28 @@ public function moderate_coursework_submission_directly(
+     *
+     * @Then /^I should( not)? see a submit button "(?P<label>[^"]*)"$/
+     */
+    public function should_see_submit_button($not, $label) {
```
Do we need a custom step for this? Can't we just use core's `And "Save changes" "button" should not exist`?

opitz replied:
refactored.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -8,13 +8,50 @@ Feature: Double marking - blind
-      | courseworkmarker    | courseworkmarker    |           |
-      | norole              | norole              |           |
+      | shortname             | name                  | archetype |
+      | courseworkexamoffice  | courseworkexamoffice  |           |
+      | courseworkmarker      | courseworkmarker      |           |
+      | courseworkmoderator   | courseworkmoderator   |           |
```
Could we drop the "coursework" prefix? It's hopefully obvious from the context.

```
    And the following "roles" exist:
      | shortname  | name       | archetype |
      | examoffice | examoffice |           |
      | marker     | marker     |           |
      | moderator  | moderator  |           |
      | norole     | norole     |           |          
```

opitz replied:
I am just copying the role shortnames used by UCL (https://workflows.preview-moodle.ucl.ac.uk/admin/roles/manage.php), so keeping them.
Refactored 'courseworkmarker' to 'courseworkdbm'

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3963,4 +3963,131 @@ public function role_has_capabilities($rolename, \Behat\Gherkin\Node\TableNode $
+            $record = new stdClass();
+            $record->roleid      = $fromrole->id;
+            $record->allowassign = $torole->id;
+            $DB->insert_record('role_allow_assign', $record);
```
Could just have:

```
            core_role_set_assign_allowed($fromrole->id, $torole->id);
```

opitz replied:
Nice! refactored.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -30,7 +30,6 @@ Feature: Double marking - blind
     And I am on the "Course 1" "course" page logged in as "admin"
     And I add a coursework activity to course "Course 1" section "2" and I fill the form with:
       | Coursework title                                            | Coursework – Double marking blind                 |
-      | Formative or summative?                                     | Summative - counts towards the final module mark  |
```
I know it's a pain keeping the tables lined up but we can remove lots of spaces now:

```
      | Coursework title                                            | Coursework – Double marking blind |
      | Description                                                 | Test coursework description       |
      | Display description on course page                          | Yes                               |
      | Start date                                                  | ##now##                           |
      | Deadline for submissions:                                   | ##now + 15 minutes##              |
      | Use marking deadline                                        | Yes                               |
      | Types of file that students are allowed to submit           | pdf                               |
      | Enable plagiarism flagging                                  | Yes                               |
      | Number of times each submission should initially be marked. | 2                                 |
      | Marker allocation enabled                                   | Yes                               |
      | Marker allocation strategy                                  | Manual                            |
      | Automatic agreement of marks                                | Percentage distance               |
      | Automatic agreement range                                   | 10                                |
      | View initial markers' grades                                | No                                |
      | Auto-populate agreed feedback comment                       | Yes                               |
      | Blind marking                                               | Yes                               |
      | Marker anonymity                                            | Yes                               |
```

opitz replied:
blanks removed

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -57,7 +57,7 @@ Feature: Double marking - blind
     Then I should see "Coursework – Double marking blind"
 
-  Scenario: Add assessors
+  Scenario: Add markers
```
IMO this test is unnecessary. The _Add markers_ link takes you to the standard Moodle activity's _Permissions_ > _Locally assigned roles_ page. So this scenario is only testing core functionality, not Coursework.

opitz replied:
This is a test from Eliots test sheet - but I agree that it isn't testing anything related to coursework, so removed.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -0,0 +1,93 @@
+      | Marker anonymity                                            | Yes |
+      | Blind marking                                               | Yes |
+
+    Then I should see "Coursework – Double marking blind"
```
What is this testing? That the activity can be added? We could do that with:

```
  Scenario: Create coursework assignment with double marking
    Given there is a double-blind marking coursework
    And I am on the "Course 1" "course" page logged in as "manager"
    Then I should see "Coursework 1"
```

If it's testing something specific in the form we could state that under `Scenario:`.

Unless it's testing something specific to blind double-marking it doesn't seem to belong in this `.feature` file.

opitz replied:
From Eliots test sheet. It supposed to test the complete manual setup of a double blind marking. If you insist of having it removed or changed I am open to persuasion...

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -29,19 +66,39 @@ Feature: Double marking - blind
-      | student2  | Student   | 2        | student2@example.com |
-      | student3  | Student   | 3        | student3@example.com |
+      | username    | firstname   | lastname | email                  |
+      | teacher1    | Teacher     | 1        | teacher1@example.com   |
```
I can't see this user is used anywhere.

opitz replied:
removed

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -0,0 +1,93 @@
+      | CLC  | core_course | course | 0      |
+    And the following "custom fields" exist:
+      | name        | shortname   | category | type |
+      | Course Year | course_year | CLC      | text |
```
Is this custom field needed? I can't see that it's part of any of the tests.

opitz replied:
You are right, removed.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -316,24 +365,23 @@ Feature: Double marking - blind
+      | student2  | C1     | student              |
+      | student3  | C1     | student              |
+
+    And I am on the "Course 1" "course" page logged in as "manager"
```
```
378     And I am on the "Course 1" "course" page logged in as "manager"
        ⋮
393     And I log out
```

The steps between logging in and logging out all make changes directly in the database so we don't need to log in as a manager here.

opitz replied:
removed

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -106,28 +83,19 @@ Feature: Double marking - blind
+    And there is a double-blind marking coursework
+    And the following "course enrolments" exist:
+      | user      | course | role             |
+      | marker1   | C1     | teacher          |
```
Can't we have a single `And the following "course enrolments" exist:` in `Background:`? We create the users there we might as well enrol them too. It saves us 104 lines of repetition.

opitz replied:
Good idea, but I experienced weird behaviour: when users are enrolled in the background and a coursework is added in a scenario (e.g. "Mark the assignments") _after_ the users are enrolled I get a random(?) assignments of markers - even if explicitly setting markers in the scenario. Since I don't know (yet) what causes this I keep things as is for now.

This is what I found out so far: 
- When I create a coursework _before_ adding course enrolments I can allocate markers as expected.
- When a coursework is created _after_ adding course enrolments, allocating markers in the scenario is ignored and a random(?) setting for markers is allocated (see screenshot).
No idea what happens here...

<img width="1167" height="513" alt="Screenshot 2026-01-09 at 10 25 44" src="https://github.com/user-attachments/assets/65ce216e-de34-48ca-9281-34a72a162eca" />

leonstr replied:
See [this comment](https://github.com/ucl-isd/moodle-mod_coursework/pull/181#discussion_r2637935184) re setting `$coursework->assessorallocationstrategy`.

opitz replied:
learned something here...

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3823,4 +3813,73 @@ public function set_extension_for_user($fullname, $cwname, $datestr) {
+            'allocatableuser' => $student->id,
+            'stageidentifier' => 'assessor_' . $assignnr,
+            'allocatabletype' => 'user',
+        ]);
```
If we're doing the assignments in Behat test why do we need to check if there are existing assignments? Instead we can just add the required assignments only.

I think this just needs to be:

```
        ⋮
        $record = new stdClass();
        $record->courseworkid = $cw->id;
        $record->allocatableid = $student->id;
        $record->allocatableuser = $student->id;
        $record->assessorid = $marker->id;
        $record->stageidentifier = $stageidentifier;
        $record->allocatabletype = 'user';
        $record->ismanual = 1;

        $generator = testing_util::get_data_generator()->get_plugin_generator('mod_coursework');
        $allocation = $generator->create_allocation($record);
```

(See also the note for `there_is_a_double_blind_marking_coursework()` to fix unexpected allocations).

opitz replied:
refactored i_assign_user_as_role_for_student_in_coursework()

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3823,4 +3813,73 @@ public function set_extension_for_user($fullname, $cwname, $datestr) {
+            $DB->update_record('coursework_allocation_pairs', $record);
+        } else {
+            $DB->insert_record('coursework_allocation_pairs', $record);
+        }
```
Allocations are cached by mod_coursework so adding these directly in the database may cause unexpected test failures. We should purge the cache after making changes:

```
        \mod_coursework\models\allocation::remove_cache($this->coursework->id);
```

opitz replied:
thanks!

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3823,4 +3813,73 @@ public function set_extension_for_user($fullname, $cwname, $datestr) {
+        global $DB;
+
+        // Find the coursework by name.
+        $cw = $DB->get_record('coursework', ['name' => $cwname], '*', MUST_EXIST);
```
We can use `$this->coursework->id`.

opitz replied:
refactored.

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1619,6 +1619,35 @@ public function there_is_a_double_blind_marking_coursework() {
+         */
+        $generator = testing_util::get_data_generator()->get_plugin_generator('mod_coursework');
+
+        $coursework = new stdClass();
```
We need to set `$coursework->assessorallocationstrategy = "none"`. When adding this in Moodle this is the default, but if we add this to the database directly the default is `'equal'` which causes `mdl_coursework_allocation_pairs` to be automatically populated if there are already users enrolled.

opitz replied:
Ahh! This explains the "random" assignments of markers to a coursework set up _after_ the course enrolment. 

But there is still an issue: it seems that when a coursework is created _after_ course enrolments any marker setting in the scenario is ignored (instead of "random" markers I end up with _no_ markers now...). 

Applying this anyway (but keeping the course enrolment after the coursework setup for now...).

leonstr replied:
> But there is still an issue: it seems that when a coursework is created after course enrolments any marker setting in the scenario is ignored (instead of "random" markers I end up with no markers now...).

See [this comment](https://github.com/ucl-isd/moodle-mod_coursework/pull/181#discussion_r2637932709) re the need to clear the cache when allocating markers directly in the database.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -833,3 +833,65 @@ Feature: Double marking - blind
+    And I follow "Agreed" in row "1"
+    Then I should see "Moderation for "
+    And I should see "Moderator 1"
+    Then I should not see a submit button "Save changes"
```
If we check something isn't present we should check that it does appear at some earlier point. Otherwise it might end up with a check for something that never appears.

opitz replied:
Good point! Amended to check moderator can edit before marks have been released.

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -174,3 +174,58 @@ Feature: Double marking - blind
     And I should not see "Student 3"
+
+  @javascript
+  Scenario: Add extension to a student
```
As Copilot has already pointed out this, test will fail after 08:00 on 1 January 2027 (375 days time), because extension dates must be in the future.

opitz replied:
Refactored.

leonstr reviewed (changes requested):
CI checks for this branch currently fail so this at least needs resolving. If any of the current Behat failures are due to bugs in Coursework then those failing scenarios should be moved to pull requests fixing these bugs.

opitz commented on tests/behat/double_marking_blind.feature:
```diff
@@ -0,0 +1,831 @@
+    And I click on "Save" "button" in the "Extended deadline" "dialogue"
+    And I should see "1 January 2027, 8:00 AM" in the "Student 1" "table_row"
+    Then I visit the coursework page
+    And I should see "1 January 2027, 8:00 AM" in the "Student 1" "table_row"
```
done.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3657,4 +3733,607 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        // Find user by full name (firstname + lastname).
+        if (strpos($fullname, ' ') === false) {
+            throw new Exception("Full name '{$fullname}' must contain a space separating first and last name.");
```
done

opitz committed:
refactored behat steps to use $this->coursework

opitz committed:
using dynamic dates for deadline

opitz committed:
updated full name format error message

opitz committed:
using ternary operator to determine extended deadline

opitz committed:
Refactored Exception to coding_exception

opitz committed:
refactored there_is_a_double_blind_marking_coursework() and there_is_a_blind_marking_moderation_coursework() into there_is_a_blind_marking_coursework()

opitz committed:
refactored i_assign_user_as_role_for_student_in_coursework() into i_allocate_the_following_markers()

opitz committed:
removed i_click_add_mark_for_marker_in_row()

opitz committed:
added student_has_a_finalised_submission()

opitz committed:
refactored i_follow_in_row()

opitz committed:
refactored i_should_see_text_in_row()

opitz committed:
Refactored role 'courseworkmarker' to 'courseworkdbm'

opitz committed:
using core_role_set_assign_allowed()

opitz committed:
removed i_should_see_mark_in_row()

opitz committed:
using named_student_has_a_submission()
amended named_student_has_a_submission() to allow full names ('firstname lastname') as well

opitz committed:
removed should_see_submit_button()

opitz committed:
removed Scenario: Add markers

opitz committed:
removed unused user teacher1

opitz committed:
removed custom field

opitz committed:
removed unneccessary logins,
reamed i_allocate_the_following_markers() to the_following_markers_are_allocated()

opitz committed:
added $coursework->assessorallocationstrategy = "none"

opitz committed:
revoking refactoring i_should_see_text_in_row() for now

opitz committed:
check moderator can edit before marks have been released

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3633,4 +3709,530 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        if (strpos($fullname, ' ') === false) {
+            throw new Exception("Full name '{$fullname}' must contain a space separating first and last name.");
+        }
+        [$first, $last] = explode(' ', $fullname, 2);
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3633,4 +3709,530 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+                ]
+            )
+        ) {
+            throw new \moodle_exception("Submission for '{$studentfullname}' not found in '{$courseworkname}'.");
```
refactored.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3633,4 +3709,530 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+            $timestamp = $datestr;
+        } else {
+            $timestamp = strtotime($datestr);
+        }
```
refactored.

opitz committed:
using ExpectationException

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,381 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        // Normal positive check.
+        if (!$contains) {
+            throw new Exception("'{$text}' was not found in row {$rownumber}.\nRow contents: {$rowtext}");
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,381 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // If "not" was present in the step.
+        if (trim($not) === 'not') {
+            if ($contains) {
+                throw new Exception("'{$text}' text was found in row {$rownumber}.\nRow contents: {$rowtext}");
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,381 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // Get all rows.
+        $rows = $tbody->findAll('css', 'tr');
+        if (!isset($rows[$rownumber - 1])) {
+            throw new Exception("Row {$rownumber} does not exist in the submissions table.");
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,381 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // Find the table body.
+        $tbody = $page->find('css', 'table.mod-coursework-submissions-table tbody');
+        if (!$tbody) {
+            throw new Exception('Could not find coursework submissions table.');
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,387 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        // Find user by full name (firstname + lastname).
+        if (strpos($fullname, ' ') === false) {
+            throw new coding_exception("Full name '{$fullname}' must be in format 'Firstname Lastname' (space-separated).");
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,387 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        $record->courseworkid = $cw->id;
+        $record->allocatableid = $user->id;
+        $record->allocatabletype = 'user';
+        $record->extended_deadline = is_int($datestr) ? $datestr : strtotime($datestr);
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,387 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        foreach ($datahash as $allocate) {
+            $stages = ['assessor_1'];
+            $stages[] = isset($allocate['moderator']) ? 'moderator' : 'assessor_2';
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,387 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        // Check text inside the row.
+        // If "not" was present in the step.
+        if (trim($not) === 'not') {
```
refactored

opitz committed:
addressed copilot code review issues

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -2652,15 +2684,28 @@ public function i_have_a_submission() {
-            [$firstname]
-        );
+
+        // Check if the name consists of first- and lastname.
```
updated.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,394 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        if (strpos($fullname, ' ') === false) {
+            throw new coding_exception(
+                "Full name '{$fullname}' must be in format 'Firstname Lastname' and contain at least one space " .
+                "between first and last name.");
```
changed

opitz committed:
addressed copilot code review issues 2

opitz committed:
reintroduced checking int for datestring to avoid error

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,392 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        // Check text inside the row.
+        // If "not" was present in the step.
+        if (trim($not) === 'not') {
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1580,6 +1581,37 @@ public function there_is_a_coursework() {
+     *
+     * @Given /^there is a (double-)?blind marking( moderated)? coursework$/
+     */
+    public function there_is_a_blind_marking_coursework(bool $double = false, bool $moderated = false) {
```
keeping it

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1580,6 +1581,37 @@ public function there_is_a_coursework() {
+     *
+     * @Given /^there is a (double-)?blind marking( moderated)? coursework$/
+     */
+    public function there_is_a_blind_marking_coursework(bool $double = false, bool $moderated = false) {
```
keeping it

opitz committed:
addressed copilot code review issues 3

opitz committed:
refactored Scenario: Check moderation

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -2652,15 +2684,28 @@ public function i_have_a_submission() {
+        $firstname = $username[0];
+        $lastname = $username[1] ?? '';
+
+        if (!$lastname) {
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,392 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     * @Given /^I follow "(?P<linktext_string>[^"]*)" in row "(?P<rownumber>\d+)"$/
+     */
+    public function i_follow_in_row($linktext, $rownumber) {
+        $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//a[text()='$linktext']";
```
refactored.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,392 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        $element = $this->getSession()->getPage()->find('xpath', $xpath);
+
+        if (!$element) {
+            $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//button[text()='$linktext']";
```
refactored (see above)

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,392 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     * @return void
+     * @throws coding_exception
+     */
+    public function i_should_see_text_in_row($not, $text, $rownumber) {
```
refactored

opitz committed:
addressed copilot code review issues 4

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,394 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        if (strpos($fullname, ' ') === false) {
+            throw new coding_exception("Full name '{$fullname}' must contain at least one space between first and last name.");
+        }
+        [$first, $last] = explode(' ', $fullname, 2);
```
Refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,394 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        $element = $this->getSession()->getPage()->find('xpath', $xpath);
+
+        if (!$element) {
+            $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[$rownumber]//button[text()='$linktext']";
```
refactored

opitz committed:
addressed copilot code review issues 5

opitz committed:
addressed copilot code review issues 6

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -2652,15 +2684,28 @@ public function i_have_a_submission() {
+        $firstname = $username[0];
+        $lastname = $username[1] ?? '';
+
+        if (!isset($username[1])) {
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // Find the table body.
+        $tbody = $page->find('css', 'table.mod-coursework-submissions-table tbody');
+        if (!$tbody) {
+            throw new ExpectationException('Could not find coursework submissions table.', $this->getsession());
```
keeping lowercase as Moodle demands it: Method and function names should be simple English lowercase words.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // Get all rows.
+        $rows = $tbody->findAll('css', 'tr');
+        if (!isset($rows[$rownumber - 1])) {
+            throw new ExpectationException("Row {$rownumber} does not exist in the submissions table.", $this->getsession());
```
keeping lowercase as Moodle demands it: Method and function names should be simple English lowercase words.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+            if ($contains) {
+                throw new ExpectationException(
+                    "'{$text}' text was found in row {$rownumber}.\nRow contents: {$rowtext}",
+                    $this->getsession()
```
keeping lowercase as Moodle demands it: Method and function names should be simple English lowercase words.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        if (!$contains) {
+            throw new ExpectationException(
+                "'{$text}' was not found in row {$rownumber}.\nRow contents: {$rowtext}",
+                $this->getsession()
```
keeping lowercase as Moodle demands it: Method and function names should be simple English lowercase words.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        if (!$element) {
+            $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[{$rownumber}]
+            //button[normalize-space(.) = {$linktext}]";
```
unnecessary complication, makes human reading worse. keeping it.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+
+        if (!$element) {
+            $xpath = "(//table[contains(@class,'mod-coursework-submissions-table')]/tbody/tr)[{$rownumber}]
+            //button[normalize-space(.) = {$linktext}]";
```
unnecessary complication, makes human reading worse. keeping it.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1580,6 +1581,37 @@ public function there_is_a_coursework() {
+        $coursework->course = $this->course;
+        $coursework->startdate = time();
+        $coursework->deadline = strtotime('+15 minutes');
+        $coursework->numberofmarkers = ($double ? 2 : 1);
```
refactored

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,395 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        $record->courseworkid = $cw->id;
+        $record->allocatableid = $user->id;
+        $record->allocatabletype = 'user';
+        $record->extended_deadline = is_int($datestr) ? $datestr : strtotime($datestr);
```
refactored

opitz committed:
addressed copilot code review issues 7

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,406 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        $record = new stdClass();
+        $record->courseworkid = $cw->id;
+        $record->allocatableid = $user->id;
+        $record->allocatabletype = 'user';
```
removed

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1580,6 +1581,37 @@ public function there_is_a_coursework() {
+     *
+     * Example: And there is a double-blind marking coursework
+     * Example: And there is a blind marking moderated coursework
+     *
```
why not...

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -2652,15 +2684,28 @@ public function i_have_a_submission() {
+        $firstname = $nameparts[0];
+        $lastname = $nameparts[1] ?? '';
+
+        if (!isset($nameparts[1])) {
```
refactored

andrewhancox reviewed (changes requested):
As discussed, we need to be moving away from bespoke behat steps except where absolutely necessary.

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -1580,6 +1581,37 @@ public function there_is_a_coursework() {
+
+        $this->coursework = coursework::find($generator->create_instance($coursework)->id);
+    }
+
```
There's already custom behat steps to generate a coursework activity and change it's settings.

opitz replied:
Refactored. Also split into two files for double-blind and moderated marking tests.

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3361,6 +3406,25 @@ public function i_should_see_submitted_date($date) {
+     *
+     * @Given /^I should see late submitted date "(?P<date>(?:[^"]|\\")*)"$/
+     */
+    public function i_should_see_late_submitted_date($date) {
```
Just use a normal I should see step instead - add a data attribute if needed.

opitz replied:
This is actually a clone of _your_ function i_should_see_submitted_date() but with the added string component "Late" to make the match happen. So what is wrong with my approach that is right with yours?

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     *
+     * @param TableNode $allocations Students and their markers.
+     */
+    public function the_following_markers_are_allocated(TableNode $allocations) {
```
See my rebuild of the allocation table - with the refactoring done there this is now two steps of standard behat.

opitz replied:
I am not sure I can follow... 
Do you have a code example, please?

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        } else {
+            $DB->insert_record('coursework_feedbacks', $feedback);
+        }
+    }
```
Isn't it pretty easy to do this through the UI?

opitz replied:
It requires quite some steps (see below). In an early stage I had set it up through UI, but since it was used quite often it's better readable (and faster) when done as an extra step.
`   
Then I follow "Allocate markers"
    And I set the field "Allocation strategy" to "Manual"
    And I press "Apply"
    Then I should see "Please make sure markers are allocated"

    Then I set the field with xpath "//tr[contains(.,'Student 1')]//td[@class='assessor_1']//select" to "marker 1"
    And I set the field with xpath "//tr[contains(.,'Student 1')]//td[@class='assessor_2']//select" to "marker 2"
    And I set the field with xpath "//tr[contains(.,'Student 2')]//td[@class='assessor_1']//select" to "marker 1"
    And I set the field with xpath "//tr[contains(.,'Student 2')]//td[@class='assessor_2']//select" to "marker 2"
    And I set the field with xpath "//tr[contains(.,'Student 3')]//td[@class='assessor_1']//select" to "marker 1"
    And I set the field with xpath "//tr[contains(.,'Student 3')]//td[@class='assessor_2']//select" to "marker 2"
    And I press "Save"
`

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     * @return void
+     * @throws coding_exception
+     */
+    public function allow_role_to_assign_role(string $fromrolename, string $torolename) {
```
I think there's a core step to do this

opitz replied:
Yes, there is. But as not needed I removed it entirely.

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        }
+
+        // Click it.
+        $element->click();
```
Definitely possible using core steps - add a data attribute to the element if absolutely necessary.

opitz replied:
refactored to core steps

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+                $this->getsession()
+            );
+        }
+    }
```
Again - core behat can do this

opitz replied:
refactored to core steps

andrewhancox commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,405 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        } else {
+            $DB->insert_record('coursework_mod_agreements', $agreement);
+        }
+    }
```
Why not just do it through the UI?

opitz replied:
Too repetitive: I need to mark submissions 30 times for my tests. This would cause a lot more steps with no extra test gain.

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3686,406 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        // Find user by full name (firstname + lastname).
+        if (strpos($fullname, ' ') === false) {
+            throw new coding_exception("Full name '{$fullname}' must contain at least one space between first and last name.");
+        }
```
refactored

opitz committed:
refactored into two files: double_marking_blind.feature and moderation_marking.feature
using existing behat steps to replace there_is_a_blind_marking_coursework()

opitz committed:
updated message

opitz committed:
removed allow_role_to_assign_role() as no longer needed

opitz committed:
replaced i_should_see_text_in_row() with core step and removed it

opitz committed:
replaced i_follow_in_row() with core step and removed it

opitz commented on tests/behat/double_marking_blind.feature:
```diff
@@ -0,0 +1,349 @@
+    And I should see "##+ 1 month##%d %B %Y, 8:00 AM##" in the "Student 1" "table_row"
+
+  @javascript @_file_upload
+  Scenario: Student can submit a PDF fil
```
fixed

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3662,281 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     *
+     * @param string $fullname
+     * @return mixed|stdClass
+     * @throws coding_exception
```
Added

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3662,281 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+     * Return a user record from a given full name ("firstname lastname")
+     *
+     * @param string $fullname
+     * @return mixed|stdClass
```
added

opitz committed:
addressed copilot code review issues 8

opitz commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,401 @@
+    And I am on the "Course 1" "course" page logged in as "moderator1"
+    And I follow "Coursework 1"
+    And I click on "Agreed" "link" in the ".mod-coursework-submissions-row:nth-child(1)" "css_element"
+    Then I should see "Moderation for "
```
This is expected

opitz commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -3622,4 +3662,281 @@ public function i_set_the_field_to_replacing_line_breaks($field, $value) {
+        global $DB;
+
+        // Check date string.
+        if (is_int($datestr) || (is_string($datestr) && ctype_digit($datestr))) {
```
Test will fail without. Will leave it in.

opitz commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,401 @@
+    And I am on the "Course 1" "course" page logged in as "moderator1"
+    And I follow "Coursework 1"
+    And I click on "Agreed" "link" in the ".mod-coursework-submissions-row:nth-child(1)" "css_element"
+    Then I should see "Moderation for "
```
This is expected.

opitz committed:
using unique feature

opitz committed:
removed trailing spaces

opitz commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,401 @@
+    And I am on the "Course 1" "course" page logged in as "moderator1"
+    And I follow "Coursework 1"
+    And I click on "Agreed" "link" in the ".mod-coursework-submissions-row:nth-child(1)" "css_element"
+    Then I should see "Moderation for "
```
removed

leonstr commented on tests/behat/behat_mod_coursework.php:
```diff
@@ -2681,15 +2681,28 @@ public function i_have_a_submission() {
+                [$firstname]
+            );
+        } else {
+            $userid = $DB->get_field_sql(
+                "SELECT id FROM {user} WHERE firstname = ? AND lastname = ?",
+                [$firstname, $lastname]
+            );
```
Maybe:

```
            $userid = $DB->get_field('user', 'id', ['firstname' => $firsname, 'lastname' => $lastname]);
```

leonstr commented on tests/behat/double_marking_blind.feature:
`Background` should start `Given` followed by `And`s, i.e.:

```
  8     Given the following "roles" exist:
        ⋮
 83     And there is a course
        ⋮
 96     And there is a coursework
```

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -229,3 +229,73 @@ Feature: Double marking - blind
+    # Student has extension so late submission is in time
+    When I am on the "Course 1" "course" page logged in as "student1"
+    And I follow "Coursework 1"
+    When I visit the coursework page
```
Should be `And` not `When`?

leonstr commented on tests/behat/double_marking_blind.feature:
```diff
@@ -372,3 +420,31 @@ Feature: Double marking - blind
+    And I follow "Coursework 1"
+    Then I should see "Submission"
+    And I should see "In marking"
+    And I should not see "Edit your submission"
```
Could use `But` not `And`.

leonstr commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,404 @@
+  I want to perform the full coursework workflow with blind marking.
+
+  Background:
+    And the following "roles" exist:
```
Background should start `Given` followed by `And`s, i.e.:

```
  7   Background:
  8     Given the following "roles" exist:
        ⋮
 83     And there is a course   
        ⋮
 96     And there is a coursework
```

leonstr commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,404 @@
+    Then I should see "Disagreed" in row "2"
+    When I follow "Disagreed"
+    And I wait until the page is ready
+    Then I should see "I don't like it!"
```
Should be `And` not `Then`.

leonstr commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,404 @@
+      | Comment   | I don't like it at all! |
+
+    And I am on the "Course 1" "course" page logged in as "marker1"
+    And I follow "Coursework 1"
```
`When`?

leonstr commented on tests/behat/moderation_marking.feature:
```diff
@@ -0,0 +1,404 @@
+    And I follow "Agreed" in row "1"
+    Then I should see "Moderation for "
+    And I should see "Moderator 1"
+    And "Save changes" "button" should not exist
```
Could use `But`.

leonstr reviewed (commented):
Some nitpicks.
